source/material_model/drucker_prager_compositions.cc

```diff
@@ -0,0 +1,733 @@
+
+      // sum the compositional fields for normalization purposes
+      double sum_composition = 0.0;
+      for ( unsigned int i=0; i < x_comp.size(); ++i)
+        sum_composition += x_comp[i];
+
+      if (sum_composition >= 1.0)
+        {
+          volume_fractions[0] = 0.0;  // background material
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1]/sum_composition;
+        }
+      else
+        {
+          volume_fractions[0] = 1.0 - sum_composition; // background material
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1];
+        }
+      return volume_fractions;
+    }
+
+    template <int dim>
+    double
+    DruckerPragerCompositions<dim>::
+    compute_second_invariant(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const
+    {
+
+      const double edot_ii_strict = /*std::sqrt(0.5) */ std::sqrt(strain_rate*strain_rate); // V6
+      //if(trace(strain_rate)!=0)
+      //std::cout << "Strain rate = " << strain_rate << ", dev sr = " << deviator(strain_rate) << ", trace sr = " << trace(strain_rate) << ", trace dev sr = " << trace(deviator(strain_rate)) <<  std::endl;
+      //const double edot_ii_strict = std::sqrt(deviator(strain_rate)*deviator(strain_rate)); // V7
+      //const double edot_ii_strict = std::sqrt(std::fabs(second_invariant(deviator(strain_rate)))); // V8
+      //const double edot_ii_strict = std::sqrt(std::fabs(second_invariant(strain_rate))); // V9//std::sqrt(deviator(in.strain_rate[i])*deviator(in.strain_rate[i]));//2 * std::sqrt(0.5*deviator(in.strain_rate[i])*deviator(in.strain_rate[i]));
+      const double edot_ii = std::max(edot_ii_strict, min_strain_rate*min_strain_rate);
+      return edot_ii;
+    }
+
+    template <int dim>
+    double
+    DruckerPragerCompositions<dim>::
+    compute_viscosity(const double edot_ii,const double pressure,const int comp,const double prefactor,const double alpha, const double eref, const double min_visc, const double max_visc) const
+    {
+      double viscosity;
+      if (comp == 0)
+        {
+
+          const double strength = ( (dim==3)
+                                    ?
+                                    ( 6.0 * cohesion[comp] * std::cos(phi[comp] * numbers::PI/180) + 6.0 * std::max(pressure,0.0) * std::sin(phi[comp] * numbers::PI/180) )
+                                    / ( std::sqrt(3.0) * ( 3.0 + std::sin(phi[comp]) ) )
+                                    :
+                                    cohesion[comp] * std::cos(phi[comp] * numbers::PI/180) + std::max(pressure,0.0) * std::sin(phi[comp] * numbers::PI/180) );
+
+          // Rescale the viscosity back onto the yield surface
+          if (strength != 0 && edot_ii != 0)
+            viscosity = strength / ( 2.0 * std::sqrt(0.5) * edot_ii );
+          else
+            viscosity = ref_visc;
+
+          viscosity =ref_visc*viscosity/(ref_visc+viscosity);
```

Comment:
Found the problem. Forgot that this term was also present in the Spiegelman paper, and should also present in the derivative. This explains why the analytical derivative didn't converge well, while the finite difference did.

Reply:
OK, great. So everything now agrees?

Reply:
It only agrees when I use 
```
 0    1/2
1/2    0
```
And only when some parts of the strain-rate are not zero, because when they are the analytical derivatives in that part of the tensor becomes zero, and the finite differences still has a value other then zero.